Update how to create a module guide (#1649)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/build-blockchain/module/command.adoc

interface Params {
}

export class CreateHelloCommand extends BaseCommand {
	public schema = {
		$id: 'CreateHelloCommand',
		type: 'object',
		properties: {},
	};

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(_context: CommandVerifyContext<Params>): Promise<VerificationResult> {
		return { status: VerifyStatus.OK };
	}

	public async execute(_context: CommandExecuteContext<Params>): Promise<void> {
	}
}
----

== Command params & schema

The command parameters are data that is provided by the transaction, that is required by the command to execute its' business logic.
The parameters interface and schema define the data type, order of the command.

The command schema can also define additional properties like min and max length of a parameter.

For creating a Hello message, define the parameters like so:

[source,typescript]
----
interface Params {
	message: string;
}
----

The only property needed by the module is the `message` that the sender posted.

For the corresponding schema, create a new file `schema.ts` in the root folder of the Hello module.

This file will be used to store all schemas that the module requires, for a better overview.

> .schema.ts

Reviewer: I think it is important to use full file paths Like on [Line 133] or at least the relevant module directory. `hello/schema.ts`

**P.S. This applies to all guides**

Then, call the method at the bottom of the `init()` method of the module and use the respective config values as parameters:

.hello/module.ts
[source,typescript]
----
// [...]
export class HelloModule extends BaseModule {
    // [...]
    // eslint-disable-next-line @typescript-eslint/require-await
    public async init(blacklist: string[]): Promise<void> {
        // [...]
        // Call the command init() method with config as parameter
        this.commands[0].init(config).catch(err => {
            console.log("Error: ", err);
        });
    }
    // [...]
}
----

Now, the `blacklist`, `minMessageLength` and `maxMessageLength`, which are defined in the config.json file, are available in the command, and we can move on to implement the <<command-verification>>.

== Command verification

The command is always verified, before it is executed by the node like defined in the <<command-execution>>.
The verification of the command is defined in the `verify()` method of the command.

> There `CreateHello` command expects only one single parameter in the transaction, and this is the Hello message.

Author: ```suggestion
The `CreateHello` command expects only one single parameter inside of the transaction, and this is the Hello message.
```

. the message should not contain any of the words defined in the blacklist of module config.

We don't need to validate point 1. and 2. in the `verify()` method, because it is already validated by the schema.

Point 3. however, the blacklisted words, cannot be checked with the schema.
So let's implement the `verify()` method to filter the message for words in the blacklist, and throw an error if any word is found.

.hello/commands/create_hello_command.ts
[source,typescript]
----
// [...]
export class CreateHelloCommand extends BaseCommand {
    public schema = createHelloSchema;
    private _blacklist!: string[];
    // [...]
    // eslint-disable-next-line @typescript-eslint/require-await
    public async verify(context: CommandVerifyContext<Params>): Promise<VerificationResult> {
        let validation: VerificationResult;
        const wordList = context.params.message.split(" ");
        const found = this._blacklist.filter(value => wordList.includes(value));
        if (found.length > 0) {
            validation = {
                status: VerifyStatus.FAIL,
                error: new Error(
                    `Illegal word in hello message: ${  found.toString()}`

>                 )

Reviewer 2: for the fail case, you can just throw error for simplicity

Author: Can you give an example? Not sure I understand

Reviewer 2: instead of return
```
status: VerifyStatus.FAIL,
                error: new Error(
                    `Illegal word in hello message: ${  found.toString()}`
                )
```
you can do
```
throw new Error(
                    `Illegal word in hello message: ${  found.toString()}`
                )
```
and framework will handle internally in the same way

            };
        } else {
            validation = {
                status: VerifyStatus.OK
            };
        }

>         context.logger.info(validation,"validation");

Reviewer 2: validation log should exist, so maybe not needed here

Author: no, i don't think it exists currently, that is why i added it here

Reviewer 2: even in the debug mode?

        await messageSubstore.set(context, senderAddress, {
            message: context.params.message,
        });

        // 4. Get the Hello counter from the counter store.
        const helloBuffer = Buffer.from('hello','utf8');
        let helloCounter: CounterStoreData;
        try {
            helloCounter = await counterSubstore.get(context, helloBuffer);
        } catch (error) {
            helloCounter = {
                counter: 0,
            }
        }
        // 5. Increment the Hello counter +1.
        helloCounter.counter+=1;

        // 6. Save the Hello counter to the counter store.
        await counterSubstore.set(context, helloBuffer, helloCounter);
	}
}
----

== Try the new command out

As a final step, let's try out the command that we just created, by posting a "Create Hello" transaction to the node.

In the root folder of the Hello client, execute the following steps in the terminal:

. Rebuild the client:

  npm run build

. Start the client:

 ./bin/run start --config=config/custom_config.json

. In another terminal window, create the transaction:

>  % ./bin/run transaction:create hello createHello 10000000 --params='{"message":"Hello Lisk SDK v6!"}' --json --key-derivation-path="m/44'/134'/0'" --pretty

Reviewer 2: `--key-derivation-path="m/44'/134'/0'"` this is default derivation path?

Author: yes, i think so

Reviewer 2: in case of `default` derivation path, i think you don't need to specify

Author: updated the command and added a TIP explaining default key derivaiton path

Also, although the counter is created and incremented, however, there is no way for an external service to request the data.

To get Hello messages and the counter from the module, implement xref:{url_guides_module_endpoints_methods}[Endpoints and Methods] as explained in the next chapter.

=== Testing the command verification

To verify, if the verification of the command works as expected, create a transaction, similar to how it is done in the previous section <<try-the-new-command-out>>.

But in this case, we want the transaction to be invalid, to verify it is using the values defined in the `custom_config.json`, that we created in guide xref:{url_guides_module_config}[].
Therefore, it should violate at least one of the three command validations:

. Minimum Hello message length: 5.
. Maximum Hello message length: 300.
. The Hello message contains none of the blacklisted words.

For example, create the following Hello trasnaction, which is violating the third requirement by including a blacklisted word:

  % ./bin/run transaction:create hello createHello 10000000 --params='{"message":"Hello this is an illegalWord1"}' --json --pretty

> Then, <<send-transaction,send the transaction>> to the node, and wait for the response.

Author: yes you are right, somehow the ID gets not created. I moved it a little, and it seems to work now

<3> Will contain unit tests for the command.
Currently, it contains the auto-generated test skeletons.

Additionally, it will already import and register the new command in the Hello module:

.hello_client/src/app/modules/hello/module.ts
[source,typescript]
----
import { CreateHelloCommand } from "./commands/create_hello_command";
// [...]
export class HelloModule extends BaseModule {
    // [...]
	public commands = [new CreateHelloCommand(this.stores, this.events)];
    // [...]
}
----

== Command class & skeleton

The command `lisk generate:command` already created the class `CreateHelloCommand` which contains skeletons for the most important components of the command.

The command class always extends from the `BaseCommand`, which is imported from the `lisk-sdk` package.

However, this command is not performing any functions yet.
To change this, we implement the methods of the command in the following chapters.

Open the command skeleton in `create_hello_command.ts`:

.hello_client/src/app/modules/hello/commands/create_hello_command.ts
[source,typescript]
----
import {
    BaseCommand,
    CommandVerifyContext,
    CommandExecuteContext,
	VerificationResult,

> 	VerifyStatus,

Reviewer 2: Indentation problem

    public schema = createHelloSchema;
	private _blacklist!: string[];

    public async init(config: ModuleConfig): Promise<void> {
		// Set _blacklist to the value of the blacklist defined in the module config
		this._blacklist = config.blacklist;
		// Set the max message length to the value defined in the module config
		this.schema.properties.message.maxLength = config.maxMessageLength;
		// Set the min message length to the value defined in the module config
		this.schema.properties.message.minLength = config.minMessageLength;
	}
    // [...]
}
----

To store the blacklisted words from the module config in the command, create a new private command attribute `_blacklist`.
Inside the `init()` method of the command, assign the blacklist defined in the module config to `this._blacklist`, and also update the command schema with the minimum and maximum message length values defined in the config.

Then, call the method at the bottom of the `init()` method of the module and use the respective config values as parameters:

.hello_client/src/app/modules/hello/module.ts
[source,typescript]
----
// [...]
export class HelloModule extends BaseModule {
    // [...]

>     public async init(blacklist: string[]): Promise<void> {

Reviewer: ```suggestion
    public async init(args: ModuleInitArgs): Promise<void> {
```

        // Get the module config defined in the config.json file
		const { moduleConfig } = args;
		// Overwrite the default module config with values from config.json, if set
		const config = utils.objects.mergeDeep({}, defaultConfig, moduleConfig) as ModuleConfigJSON;
		// Validate the provided config with the config schema
		validator.validate<ModuleConfigJSON>(configSchema, config);
        // Call the command init() method with config as parameter
        this.commands[0].init(config).catch(err => {
            console.log("Error: ", err);

>         });

Reviewer 2: indentation problem

. Save the Hello message to the message store, using the `senderAddress` as the key, and the `message` as the value.
. Get the Hello counter from the counter store.
. Increment the Hello counter +1.
. Save the Hello counter to the counter store.
. Emit a "New Hello" event.

The corresponding code is shown below:

.hello_client/src/app/modules/hello/commands/create_hello_command.ts
[source,typescript]
----
import {
	BaseCommand,
	CommandVerifyContext,
	CommandExecuteContext,
	VerificationResult,
	VerifyStatus,
} from 'lisk-sdk';
import { createHelloSchema } from '../schema';
import { MessageStore } from '../stores/message';
import { counterKey, CounterStore, CounterStoreData } from '../stores/counter';
import { ModuleConfig } from '../types';
import { NewHelloEvent } from '../events/new_hello';

export class CreateHelloCommand extends BaseCommand {
    public schema = createHelloSchema;
    private _blacklist!: string[];
    public async init(config: ModuleConfig): Promise<void> {
		// [...]
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(context: CommandVerifyContext<Params>): Promise<VerificationResult> {
		// [...]

> 	}

Reviewer 2: indentation problem
